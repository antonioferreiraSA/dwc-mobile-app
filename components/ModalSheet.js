/**
 * Modal sheet that slides up from the bottom,
 * similar to Facebook's webviews
 *
 * - includes the button that triggers showing the modal sheet
 * - only requires the button title and children to render
 */

import React, { useState, useEffect } from 'react';
import {
  KeyboardAvoidingView,
  Platform,
  ScrollView,
  StyleSheet,
  View,
} from 'react-native';
import Modal from 'react-native-modal';
import Button from './shared/Button';
import Colors from '../constants/Colors';

export default ({
  success = false,
  buttonTitle = 'ðŸš¨ needs a title ðŸš¨',
  handleOpenModal,
  children,
}) => {
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    // only close this modal if it's already visible and
    // we've successfully finished our flow or whatevs
    if (isVisible) {
      setIsVisible(!success);
    }
  }, [isVisible, success]);

  return (
    <>
      <Button
        title={buttonTitle}
        style={styles.button}
        onPress={() => {
          if (handleOpenModal) {
            handleOpenModal();
          }
          setIsVisible(true);
        }}
      />
      <Modal
        isVisible={isVisible}
        onBackButtonPress={() => setIsVisible(false)}
        onBackdropPress={() => setIsVisible(false)}
        onSwipeComplete={() => setIsVisible(false)}
        swipeDirection="down"
        style={styles.modal}
      >
        <KeyboardAvoidingView
          behavior={Platform.OS == 'ios' ? 'padding' : 'height'}
          style={styles.container}
        >
          <View style={styles.dragBar} />
          <ScrollView contentContainerStyle={styles.contentContainer}>
            {children}
          </ScrollView>
        </KeyboardAvoidingView>
      </Modal>
    </>
  );
};

const dragBarColor = 'rgba(255,255,255,0.3)';

const styles = StyleSheet.create({
  button: { marginBottom: 16 },
  modal: {
    margin: 0,
    justifyContent: 'flex-end',
  },
  closeButton: { alignSelf: 'flex-end' },
  container: {
    height: '85%',
    paddingTop: 10,
    paddingBottom: 20,
    position: 'relative',
    borderTopLeftRadius: 10,
    borderTopRightRadius: 10,
    backgroundColor: Colors.darkerGray,
  },
  // needed to allow scrolling on android
  contentContainer: { flexGrow: 1 },
  dragBar: {
    width: 100,
    height: 6,
    marginBottom: 30,
    alignSelf: 'center',
    borderRadius: 6,
    backgroundColor: dragBarColor,
  },
});
