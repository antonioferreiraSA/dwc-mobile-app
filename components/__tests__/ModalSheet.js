import React from 'react';
import { View, Text } from 'react-native';
import {
  flushMicrotasksQueue,
  render,
  fireEvent,
} from '@testing-library/react-native';
import ModalSheet from '../ModalSheet';

describe('<ModalSheet/>', () => {
  test('render the modal sheet with defaults', async () => {
    const { toJSON } = render(<ModalSheet />);

    expect(toJSON()).toMatchSnapshot();
  });

  test.skip('render the modal sheet and close on success', async () => {
    const { update, toJSON, getByText } = render(
      <ModalSheet buttonTitle="Join Group">
        <View>
          <Text>Join Echo Group</Text>
        </View>
      </ModalSheet>
    );

    expect(toJSON()).toMatchSnapshot();

    fireEvent.press(getByText('Join Group'));

    expect(toJSON()).toMatchSnapshot();

    // update ModalSheet with `success` prop
    update(
      <ModalSheet buttonTitle="Join Group" success>
        <View>
          <Text>Join Echo Group</Text>
        </View>
      </ModalSheet>
    );

    // wait for the modal to close
    await flushMicrotasksQueue();

    expect(toJSON()).toMatchSnapshot();
  });
});
