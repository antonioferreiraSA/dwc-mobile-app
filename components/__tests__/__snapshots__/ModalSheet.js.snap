// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalSheet/> render the modal sheet and close on success 1`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#313d48",
          "borderRadius": 6,
          "minHeight": 40,
          "paddingHorizontal": 24,
          "paddingVertical": 16,
        },
        Object {
          "marginBottom": 16,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 18,
              "textAlign": "center",
            },
            Object {},
          ]
        }
      >
        Join Group
      </Text>
    </View>
  </View>,
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "left": 0,
          "margin": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
    >
      <View
        style={
          Object {
            "backgroundColor": "#212931",
            "borderTopLeftRadius": 10,
            "borderTopRightRadius": 10,
            "height": "85%",
            "paddingBottom": 20,
            "paddingHorizontal": 16,
            "paddingTop": 10,
            "position": "relative",
          }
        }
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "rgba(255,255,255,0.3)",
              "borderRadius": 6,
              "height": 6,
              "marginBottom": 30,
              "width": 100,
            }
          }
        />
        <RCTScrollView
          contentContainerStyle={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <View>
            <View>
              <Text>
                Join Echo Group
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>,
]
`;

exports[`<ModalSheet/> render the modal sheet and close on success 2`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#313d48",
          "borderRadius": 6,
          "minHeight": 40,
          "paddingHorizontal": 24,
          "paddingVertical": 16,
        },
        Object {
          "marginBottom": 16,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 18,
              "textAlign": "center",
            },
            Object {},
          ]
        }
      >
        Join Group
      </Text>
    </View>
  </View>,
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "left": 0,
          "margin": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 1334,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
    >
      <View
        style={
          Object {
            "backgroundColor": "#212931",
            "borderTopLeftRadius": 10,
            "borderTopRightRadius": 10,
            "height": "85%",
            "paddingBottom": 20,
            "paddingHorizontal": 16,
            "paddingTop": 10,
            "position": "relative",
          }
        }
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "rgba(255,255,255,0.3)",
              "borderRadius": 6,
              "height": 6,
              "marginBottom": 30,
              "width": 100,
            }
          }
        />
        <RCTScrollView
          contentContainerStyle={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <View>
            <View>
              <Text>
                Join Echo Group
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>,
]
`;

exports[`<ModalSheet/> render the modal sheet and close on success 3`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#313d48",
          "borderRadius": 6,
          "minHeight": 40,
          "paddingHorizontal": 24,
          "paddingVertical": 16,
        },
        Object {
          "marginBottom": 16,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 18,
              "textAlign": "center",
            },
            Object {},
          ]
        }
      >
        Join Group
      </Text>
    </View>
  </View>,
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "left": 0,
          "margin": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 1334,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
    >
      <View
        style={
          Object {
            "backgroundColor": "#212931",
            "borderTopLeftRadius": 10,
            "borderTopRightRadius": 10,
            "height": "85%",
            "paddingBottom": 20,
            "paddingHorizontal": 16,
            "paddingTop": 10,
            "position": "relative",
          }
        }
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "rgba(255,255,255,0.3)",
              "borderRadius": 6,
              "height": 6,
              "marginBottom": 30,
              "width": 100,
            }
          }
        />
        <RCTScrollView
          contentContainerStyle={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <View>
            <View>
              <Text>
                Join Echo Group
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>,
]
`;

exports[`<ModalSheet/> render the modal sheet with defaults 1`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#313d48",
          "borderRadius": 30,
          "justifyContent": "center",
          "minHeight": 60,
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        },
        Object {
          "marginBottom": 16,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        selectable={true}
        style={
          Array [
            Object {
              "color": "#cbd2d6",
              "fontFamily": "NunitoSans-Regular",
              "fontSize": 18,
              "lineHeight": 24,
              "textAlign": "auto",
            },
            Array [
              Object {
                "color": "#fff",
                "textAlign": "center",
              },
              Object {},
            ],
          ]
        }
      >
        🚨 needs a title 🚨
      </Text>
    </View>
  </View>,
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      deviceHeight={null}
      deviceWidth={null}
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      panResponderThreshold={4}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "left": 0,
          "margin": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
    >
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "#212931",
              "borderTopLeftRadius": 10,
              "borderTopRightRadius": 10,
              "height": "85%",
              "paddingBottom": 20,
              "paddingTop": 10,
              "position": "relative",
            },
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "rgba(255,255,255,0.3)",
              "borderRadius": 6,
              "height": 6,
              "marginBottom": 30,
              "width": 100,
            }
          }
        />
        <RCTScrollView
          contentContainerStyle={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <View />
        </RCTScrollView>
      </View>
    </View>
  </Modal>,
]
`;
